New Feature: Accordion RTE Plugin

stevenwoodson opened the pull request:
This is a feature that HS2 Solutions had developed allowing content authors to add an accordion to any blocks of content editable within the Rich Text Editor.

**Use**
When activated, this plugin adds a CoralUI feed icon to the full-screen editor. Clicking that icon will add the first accordion header and content area after the highest parent of the current context, this was chosen because difficulties could arise from adding within other block level elements like paragraphs or tables.
![1-new-accordion](https://cloud.githubusercontent.com/assets/1857439/17753373/2d5ece64-6495-11e6-8279-857ebca0a9d8.png)

While context is inside an existing accordion, the feed icon changes to feedAdd to signify that additional accordion items can be added. When clicked, a new item (with its own header and content) will be appended to the end of the accordion.
![2-new-item](https://cloud.githubusercontent.com/assets/1857439/17753388/4343eac0-6495-11e6-96b5-569f7def644f.png)

All items header and content areas are visible while in the full-screen editor for easier content editing, otherwise the default is for all items to start "closed" with only headers displayed. Clicking anywhere on a header will reveal its content in a slide down animation, any other items in the same accordion that are open are simultaneously closed.

**Styling**
Minimal generic styling has been included to provide the basic functionality. To negate the need for any additional dependencies, the plus sign icon beside each header (signifying that it's clickable) is the \002b unicode character. When an item is opened this plus icon is rotated 45° via CSS to become an 'X' signifying that the item can be closed.

**Preview**
![screen shot 2016-12-15 at 12 48 14 pm](https://cloud.githubusercontent.com/assets/1857439/21237476/d405f548-c2c4-11e6-98cd-332007b7652a.png)

**Installation**
I've built and tested this plugin inside the stock Geometrixx Outdoors site in AEM 6.1. To enable this plugin, via CRXDE:
- Add a new node under `/libs/foundation/components/text/dialog/items/tab1/items/text/rtePlugins` to enable the accordion addition. Name it **acs-commons** and set two properties:
  - jcr:primaryType="nt:unstructured"
  - features="rte-accordion"
- Add **acs-commons.rte-accordion** to the embed property of the site, for example in the desktop version of Geometrixx Outdoors it's located at `/etc/designs/geometrixx-outdoors/clientlibs_desktop_v1/`.

stevenwoodson commented:
Wanted to check in on this, are there any concerns with this addition?

justinedelson commented:
@schoudry can you review this?

schoudry commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/accordion/clientlibs/css/accordion.less:
```diff
@@ -0,0 +1,110 @@
+
+/* Accordion Styles */
+.rte-accordion {
```
generic style name; can it be prefixed with acs-commons; so acs-commons-rte-accordion

stevenwoodson replied:
Makes sense, just updated in a new commit

schoudry commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/accordion/rte-accordion.js:
```diff
@@ -0,0 +1,194 @@
+    };
+ 
+    // Extend toolbar to register accordion icon
+    CUI.rte.ui.cui.CuiToolbarBuilder = new Class({
```
overriding otb registration is something we wanted to avoid in acs commons (interferes with other plugin registrations); check https://github.com/Adobe-Consulting-Services/acs-aem-commons/commit/c7ded7d1033caffbac5d08afb4d5676bcc5ec847

stevenwoodson replied:
I tried removing this class and adding `tbGenerator.registerIcon(RTEAccordion.GROUP + "#" + RTEAccordion.FEATURE, "coral-Icon coral-Icon--feed");` to `initializeUI` in `RTEAccordion.TouchUIAccordionPlugin` but can't seem to get the icon to show up in the fullscreen RTE anymore.

Are there any resources/documentation online that you could point me to for this?

stevenwoodson replied:
After taking some more time to read through the whole PR for #661 and the originating PR #546 I figured out how to get this working. I've committed a change that removes this override in favor of using configuration instead.

schoudry commented on content/src/main/content/jcr_root/apps/acs-commons/authoring/accordion/rte-accordion.js:
```diff
@@ -0,0 +1,194 @@
+            if ( parent.className == RTEAccordion.FEATURE ) {
+                return parent;
+            } 
+            // We don't want to go too far up the dom, if the next parent after this is null return this node
```
too many comments :) (just an observation)

stevenwoodson replied:
Sorry, I think they're the result of my trying to understand everything that needs to happen in these classes. I haven't been able to find much in terms of documentation so this was my way of figuring it out. I can prune them down if they're hindering the readability.

schoudry commented:
@stevenwoodson cool plugin; adding to the inline comments - 

1) component rendering - i think "X" is more of a remove experience, and "-" (hyphen) a collapse experience; your thoughts?

2) had to refresh the page after editing rte content to get accordion expand working (is it just me?)

stevenwoodson commented:
Thanks again @schoudry for the review and apologies it has taken me so long to fully act on your suggestions. After taking some time to read through and understand #546 and #661, I have:

- Removed the `CUI.rte.ui.cui.CuiToolbarBuilder` override in favor of using configuration
- Moved this plugin under the `/rte-plugins/` directory to maintain the convention you started in #546 
- Changed the naming to be more consistent across category, group, HTML, and CSS
- Changed from the Plus sign icon to a more accordion suitable unicode arrow and moved it to the left side similar to what's found [here](http://semantic-ui.com/modules/accordion.html) and [here](https://jqueryui.com/accordion/).
- After some more tests I was able to reproduce your reported need to refresh before the accordion click events would work. Changed the javascript that triggers that and fixed this issue.
- Finally, now that there are a few additional steps involved in getting this activated, I added some notes to the top of the source code in a similar fashion to `dialog-plugin.js`.

I believe that covers all the concerns from your initial review, when you get some time I'd love to get some feedback on these latest changes.

stevenwoodson commented:
Hi @justinedelson & @schoudry, wanted to reach out about this to see if there are any other concerns after the edits made last year. Thank you!

schoudry commented:
apologies @stevenwoodson - due to project commitments, not finding enough free time to focus on acs commons

stevenwoodson commented:
I can definitely appreciate that @schoudry, thanks for the reply! Please let me know if there's anything I can do to make reviewing this easier for you.

stevenwoodson commented:
@schoudry  Keeping this aligned with recent changes, just merged latest `master` and modified the `.content.xml` in 93028e1 to use the GCC jsProcessor like the rest.

HitmanInWis commented:
@justinedelson @davidjgonzalez Anything we can do to help move this PR along?  The feature appears to be valued by the community (one of the winners for the AEM Community Code Contest 2017 (http://help-forums.adobe.com/content/adobeforums/en/experience-manager-forum/adobe-experience-manager.topic.html/forum__jern-hi_communitywew.html).

It appears that @schoudry may be a bit busy with other obligations to review.  Since it is an optional RTE plugin, requiring it to be "activated" in RTE configs to use, it seems pretty low risk.

schoudry commented:
@stevenwoodson @HitmanInWis quick review, looks good to me; though i would request a review by another pair of eyes @justinedelson @davidjgonzalez and a cursory run on 63 if you can (afaik acs commons UI features were not tested on 63 yet, so i guess this is not necessary)

stevenwoodson commented:
Thanks so much @schoudry, appreciate you taking the time to review. Looking forward to a final review and getting this merged! @justinedelson & @davidjgonzalez, please let me know if there's anything I can do to help with the second review on this.

stevenwoodson commented:
Coming up on the one year anniversary for this PR, if there's anything else necessary @justinedelson & @davidjgonzalez please let me know.

davidjgonzalez commented on content/src/main/content/jcr_root/apps/acs-commons/rte-plugins/accordion/.content.xml:
```diff
@@ -0,0 +1,5 @@
+<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
+    jcr:primaryType="cq:ClientLibraryFolder"
+    jsProcessor="[default:none,min:gcc]"
+    categories="rte.coralui2"/>
```
Can we change this to `categories="acs-commons.rte-accordian.rte.coralui2"` and ask consumers to create a ClientLib?

```
<jcr:root xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
   jcr:primaryType="cq:ClientLibraryFolder"
   categories="rte.coralui2"
    dependencies="[acs-commons.rte-accordian.rte.coralui2]"/>
```

I think this JS gets auto-loaded in  ".../apps/acs-commons/rte-plugins/.content.xml → ...ns/rte-plugins/dialog-plugin/.content.xml" as well - not sure why, but GH wont let me comment on that line.

Basically, we don't want to load JS/CSS unless someone explicitly enables a feature.

stevenwoodson commented:
Thank you for the response @davidjgonzalez! I've renamed the category to `acs-commons.rte-accordion.rte.coralui2` and can confirm on a fresh install that this plugin is not included by default until the installation steps are followed.

HitmanInWis commented:
Thanks for following up on this @davidjgonzalez glad to see yet another feature available for all (so we dont have to keep writing again and again :) )

davidjgonzalez commented:
@HitmanInWis sorry it took so long :(
